internal/bounce/bounce.go

```diff
@@ -145,6 +146,15 @@ func (m *Manager) runMailboxScanner() {
 
 // Record records a new bounce event given the subscriber's email or UUID.
 func (m *Manager) Record(b models.Bounce) error {
+	metrics.Counter{
```

Comment:
Instead of defining a Counter object here, IMHO a better and simpler API would be:

```go
metrics.Increment("label")
```

You can then define a `.Increment` function in package which does:

```

// Increment the counter for the corresponding key.
// This is used for Counter metric type.
func (s *Manager) Increment(label string) {
	s.metrics.GetOrCreateCounter(label).Inc()
}
```

What do you think?

Reply:
I'm not sure I understand the suggestion.

Would the idea be to pre-register all the different metric names in the Manager and give them the appropriate methods through an interface and just pass the appropriate label map when they're invoked?

Reply:
I'm re-reading and checking how this could be implemented. I don't think it's possible since every combination of metric name and label set constitutes a unique counter, and that can't be pre-declared before I know what labels will be provided

Reply:
I just gave it a try. I think it could be written, but I'm fairly certain we'll be clobbering our object and create races.

I think it could look something like this.

```go
package "metrics"
// [...]

func (c Counter) New(n string) Counter {
	return Counter{
		Name: n,
	}
}

func (c Counter) IncrementLabels(l Labels) {
	c.Labels = l
	c.Inc()
}

func (c Counter) Increment() {
	metricName := composeMetricName(c.Name, c.Labels)
	metrics.GetOrCreateCounter(metricName).Inc()
}
```

But my spider-senses are more than tingling when I look at this. :grimacing:

Reply:
Ah, sorry for not being clear in the first time. I don't think we need to register labels before hand. Here's a complete example of what I had in mind:

In the metrics internal package:

```go
// Manager represents options for storing metrics.
type Manager struct {
	metrics         *metrics.Set
	namespace       string // Optional string to prepend all label names.
}

// NewStats returns a new configured instance of Manager.
func New(ns string) *Manager {
	return &Manager{
		metrics:   metrics.NewSet(),
		namespace: ns,
	}
}

// Increment the counter for the corresponding key.
// This is used for Counter metric type.
func (s *Manager) Increment(label string) {
	s.metrics.GetOrCreateCounter(s.getFormattedLabel(label)).Inc()
}

// Decrement the counter for the corresponding key.
// This is used for Counter metric type.
func (s *Manager) Decrement(label string) {
	s.metrics.GetOrCreateCounter(s.getFormattedLabel(label)).Dec()
}

// Set updates the key with a float64 value.
// This is used for Gauge metric type.
func (s *Manager) Set(label string, val float64) {
	s.metrics.GetOrCreateFloatCounter(s.getFormattedLabel(label)).Set(val)
}

// getFormattedLabel prefixes the label with namespace (if non empty).
func (s *Manager) getFormattedLabel(l string) string {
	if s.namespace != "" {
		return s.namespace + "_" + l
	}
	return l
}
```

So, we define all the ops related to metrics exposition inside the package. This way, the API to use this package becomes cleaner.

NOTE: Since we are using `GetOrCreate` functions, we don't need to register any metrics before hand either. This makes it simpler to add ad-hoc metric names.

In the place where you need to capture a metric (i.e. bounces etc):

```go
var (
	labelBouncesProcessed         = `bounces_processed`
)
app.metrics.Increment(labelBouncesProcessed)
```

This will emit `listmonk_bounces_processed` (`listmonk` being the `namespace`) which is a valid Prometheus metric. In case you need to add labels to the metric, you can use a simple string concat `bounces_processed + "{" + label_key + "=" + label_val + "}"` (`fmt.Sprintf` can be used for readability and I don't think that should be a problem here performance wise).

Hope that makes sense. Let me know if this approach sounds easier to implement.

Reply:
Here's an example of how to add dynamic labels using string concat: https://github.com/zerodha/fastglue-metrics/blob/main/metrics.go#L98

Reply:
Yup, that makes sense to me wno. Also, that fastglue middleware gives a good exmaple of how to get request durations! I'm still early in digging into Echo middlewares. Hopefully I might be able to get that in too.

Yeah, with those examples I should be able to put it together. :pray: